src/Input/AutoResizingTextArea.js

```diff
@@ -0,0 +1,204 @@
+    onHeightChange: PropTypes.func,
+    rows: PropTypes.number,
+    rowsMax: PropTypes.number,
+    shadowClassName: PropTypes.object,
+    value: PropTypes.string,
+  };
+
+  static defaultProps = {
+    rows: 1,
+  };
+
+  static contextTypes = {
+    styleManager: customPropTypes.muiRequired,
+  };
+
+  state = {
+    height: null,
+  };
+
+  componentWillMount() {
+    // <Input> expects the components it renders to respond to 'value'
+    this.value = this.props.defaultValue;
+    this.setState({
+      height: this.props.rows * rowsHeight,
+    });
+  }
+
+  componentDidMount() {
+    this.syncHeightWithShadow();
+  }
+
+  componentWillReceiveProps(nextProps) {
+    if (nextProps.value !== this.props.value ||
+        nextProps.rowsMax !== this.props.rowsMax) {
+      this.syncHeightWithShadow(nextProps.value, null, nextProps);
+    }
+  }
+
+  handleResize = (event) => {
+    this.syncHeightWithShadow(undefined, event);
+  };
+
+  getInputNode() {
+    return this.input;
+  }
+
+  setValue(value) {
+    this.getInputNode().value = value;
+    this.syncHeightWithShadow(value);
+  }
+
+  syncHeightWithShadow(newValue, event, props) {
+    const shadow = this.shadow;
+    const singleLineShadow = this.singleLineShadow;
+
+    const hasNewValue = (newValue && newValue !== '');
+    const displayText = this.props.hintText && !hasNewValue ? this.props.hintText : newValue;
+
+    if (displayText !== undefined) {
+      shadow.value = displayText;
+    }
+
+    const lineHeight = singleLineShadow.scrollHeight;
+    let newHeight = shadow.scrollHeight;
+
+    // Guarding for jsdom, where scrollHeight isn't present.
+    // See https://github.com/tmpvar/jsdom/issues/1013
+    if (newHeight === undefined) return;
+
+    props = props || this.props;
+
+    if (props.rowsMax >= props.rows) {
+      newHeight = Math.min(props.rowsMax * lineHeight, newHeight);
+    }
+
+    newHeight = Math.max(newHeight, lineHeight);
+
+    if (this.state.height !== newHeight) {
+      this.setState({
+        height: newHeight,
+      });
+
+      if (props.onHeightChange) {
+        props.onHeightChange(event, newHeight);
+      }
+    }
+  }
+
+  handleChange = (event) => {
+    const value = event.target.value;
+    this.syncHeightWithShadow(value);
+    this.value = value;
+    if (this.props.onChange) {
+      this.props.onChange(event);
+    }
+  };
+
+  render() {
+    const {
+      onChange,
+      onHeightChange,
+      rows,
+      rowsMax,
+      hintText,
+      className,
+      ...other
+    } = this.props;
+
+    const { styleManager } = this.context;
+    const classes = styleManager.render(styleSheet);
+
+    return (
+      <div className={classnames(classes.root, className)}>
+        <EventListener target="window" onResize={this.handleResize} />
+        <textarea
+          ref={(c) => { this.singleLineShadow = c; }}
+          className={classnames(classes.shadow, classes.textarea)}
+          tabIndex="-1"
+          rows={1}
```

Comment:
rows is a `PropTypes.number` -- do you think it'd be better to make a `PropTypes.string`?